Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean

```diff
@@ -0,0 +1,298 @@
+    P y ^ 2 * P z + W.a₁ * P x * P y * P z + W.a₃ * P y * P z ^ 2
+      = P x ^ 3 + W.a₂ * P x ^ 2 * P z + W.a₄ * P x * P z ^ 2 + W.a₆ * P z ^ 3 := by
+  rw [equation, eval_polynomial, sub_eq_zero]
+
+lemma equation_zero (Y : R) : W.equation ![0, Y, 0] :=
+  (W.equation_iff ![0, Y, 0]).mpr <| by matrix_simp; ring1
+
+lemma equation_some (X Y : R) : W.equation ![X, Y, 1] ↔ W.toAffine.equation X Y := by
+  rw [equation_iff, W.toAffine.equation_iff]
+  congr! 1 <;> matrix_simp <;> ring1
+
+lemma equation_smul_iff (P : Fin 3 → R) (u : Rˣ) : W.equation (u • P) ↔ W.equation P :=
+  have (u : Rˣ) {P : Fin 3 → R} (h : W.equation P) : W.equation <| u • P := by
+    rw [equation_iff] at h ⊢
+    linear_combination (norm := (simp only [smul_fin3_ext]; ring1)) (u : R) ^ 3 * h
+  ⟨fun h => by convert this u⁻¹ h; rw [inv_smul_smul], this u⟩
+
+/-- The partial derivative $W_X(X, Y, Z)$ of $W(X, Y, Z)$ with respect to $X$.
+
+TODO: define this in terms of `MvPolynomial.pderiv`. -/
+@[pp_dot]
+noncomputable def polynomialX : MvPolynomial (Fin 3) R :=
+  C W.a₁ * X 1 * X 2 - (C 3 * X 0 ^ 2 + C (2 * W.a₂) * X 0 * X 2 + C W.a₄ * X 2 ^ 2)
+
+lemma eval_polynomialX (P : Fin 3 → R) : eval P W.polynomialX =
+    W.a₁ * P y * P z - (3 * P x ^ 2 + 2 * W.a₂ * P x * P z + W.a₄ * P z ^ 2) := by
+  rw [polynomialX]
+  eval_simp
+
+/-- The partial derivative $W_Y(X, Y, Z)$ of $W(X, Y, Z)$ with respect to $Y$.
+
+TODO: define this in terms of `MvPolynomial.pderiv`. -/
+@[pp_dot]
+noncomputable def polynomialY : MvPolynomial (Fin 3) R :=
+  C 2 * X 1 * X 2 + C W.a₁ * X 0 * X 2 + C W.a₃ * X 2 ^ 2
+
+lemma eval_polynomialY (P : Fin 3 → R) :
+    eval P W.polynomialY = 2 * P y * P z + W.a₁ * P x * P z + W.a₃ * P z ^ 2 := by
+  rw [polynomialY]
+  eval_simp
+
+/-- The partial derivative $W_Z(X, Y, Z)$ of $W(X, Y, Z)$ with respect to $Z$.
+
+TODO: define this in terms of `MvPolynomial.pderiv`. -/
+@[pp_dot]
+noncomputable def polynomialZ : MvPolynomial (Fin 3) R :=
+  X 1 ^ 2 + C W.a₁ * X 0 * X 1 + C (2 * W.a₃) * X 1 * X 2
+    - (C W.a₂ * X 0 ^ 2 + C (2 * W.a₄) * X 0 * X 2 + C (3 * W.a₆) * X 2 ^ 2)
+
+lemma eval_polynomialZ (P : Fin 3 → R) : eval P W.polynomialZ =
+    P y ^ 2 + W.a₁ * P x * P y + 2 * W.a₃ * P y * P z
+      - (W.a₂ * P x ^ 2 + 2 * W.a₄ * P x * P z + 3 * W.a₆ * P z ^ 2) := by
+  rw [polynomialZ]
+  eval_simp
+
+/-- Euler's homogeneous function theorem. -/
+theorem polynomial_relation (P : Fin 3 → R) : 3 * eval P W.polynomial =
+    P x * eval P W.polynomialX + P y * eval P W.polynomialY + P z * eval P W.polynomialZ := by
+  rw [eval_polynomial, eval_polynomialX, eval_polynomialY, eval_polynomialZ]
+  ring1
+
+/-- The proposition that a point representative $(x, y, z)$ in `W` is nonsingular.
+In other words, either $W_X(x, y, z) \ne 0$, $W_Y(x, y, z) \ne 0$, or $W_Z(x, y, z) \ne 0$. -/
+@[pp_dot]
+def nonsingular (P : Fin 3 → R) : Prop :=
+  W.equation P ∧ (eval P W.polynomialX ≠ 0 ∨ eval P W.polynomialY ≠ 0 ∨ eval P W.polynomialZ ≠ 0)
```

Comment:
Do you want `≠ 0` here (non-singular on the generic fiber) or `IsUnit` (non-singular in all fibers)?
It probably depends on how this definition will be used. `≠ 0` is certainly easier to work with (and also the correct version in some cases, e.g., when talking about integral points on elliptic curves over the rationals).

Reply:
Yes, thank you! All the suggestions here are very valid and I'll try to apply them tomorrow.

Reply:
Do you want to say one of the derivatives is a unit, or the ideal generated by the derivatives is the unit ideal? I feel the latter would generalize more broadly, e.g. if you have an elliptic curve over the integers (localized away from bad primes) then I think only the latter defines the correct version of nonsingularity. Of course over a local ring (in particular, a valuation ring) both are equivalent. I've previously [observed](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/thoughts.20on.20elliptic.20curves/near/384388729) that if you define the projective space P^n(R) as (n+1)-tuples that generate the unit ideal ⟨1⟩ quotiented by Rˣ, then `P^n(R) → P^n(Frac(R))` is a bijective if R is a Bezout domain (in particular, a PID or a valuation ring). Restricting to the tuples that generates ⟨1⟩ will be necessary to define the reduction of a point, but it could also be covered by the nonsingularity condition if we adopt the latter definition, which would have an advantage similar to [getting rid of the `v ≠ 0` layer](https://github.com/leanprover-community/mathlib4/pull/9416#discussion_r1443732520) in `Projectivization`.

Regarding the comment on integral points: my impression is that they fit more into the affine API rather than the projective API here? I'd also like to learn more bout why `≠ 0` is the correct version in that case ...

Reply:
Right - I think we want that the ideal generated by the partial derivatives is a unit. The current affine definition for nonsingularity is also using `≠ 0`, and I don't believe that that's equivalent to any of the suggestions here. For instance, I don't know how to prove `nonsingular_some` if I were to change the `nonsingular` definition.

I propose that we stick to treating `R` as if it were a field or a local ring just for this PR (and the one for Jacobian coordinates), and use `≠ 0`. We can then correct the definitions all at once in a future PR.

Reply:
I forgot - did you try to replace the affine nonsingular condition by the condition that the ideal generated by the two partial derivatives is the unit ideal, and did you encounter difficulties? That condition is equivalent to requiring all three partial derivatives generate the unit ideal, because the third derivative $F_Z(x,y,1)$ is equal to $-(xF_X(x,y,1) + yF_Y(x,y,1))$ for $[x:y:1]$ on the curve $F=0$, because of the identity $XF_X(X,Y,Z)+YF_Y(X,Y,Z)+ZF_Z(X,Y,Z)=3F(X,Y,Z)$ for a homogeneous polynomial $F$ of degree 3. (Did I post this somewhere?)

I'd like to get the definition right because to define the reduction homomorphism I do need to restrict to the tuples that generate the unit ideal. It would be enough to require one of the coordinates to be a unit to first construct the homomorphism in the case of valuation rings, but I think it would be convenient to provide the homomorphism for a more general class of rings (probably Bezout domains). It's also nice that the "generates the unit ideal" condition is "functorial" in the sense that it's preserved by arbitrary RingHoms.

It's nice that you implemented group operations in terms of projective coordinates and connected it to the affine operations in the case of a field, but with the right definition of nonsingularity, we can probably show the points form a group under the operations for a much more general class of rings (domains?). But maybe the field case already justifies merging this PR as is; I'll let other reviewers decide ...

Reply:
Yes, you posted this in the original PR. I don't think I would have any difficulties changing the affine nonsingular condition first, and then the projective and Jacobian versions, but like I said I'd like to do it in a separate PR. And yes, by changing the definition of nonsingularity we should be able to prove the group law for these domains.

Reply:
@jcommelin asked about this PR today, so let's decide now: you think it's easier to merge this PR (and the Jacobian counterpart) first, and later open a PR that changes the definition of `nonsingular` for affine, projective and Jacobian coordinates all at once? I would think it's easier to open a PR to change the affine definition, and then update the projective and Jacobian PRs after it's merged, but if you prefer that way, let's merge this now.

Reply:
Yes, I would merging this and the Jacobian counterpart now, and do the generalisation later. One reason being that I won't have time to do the new PR until April, but I would prefer to have this ready as soon as possible.